controllers/operatorpolicy_controller.go

```diff
@@ -1426,187 +1470,346 @@ func (r *OperatorPolicyReconciler) handleInstallPlan(
+			continue
 		}
 
-		for _, acceptableCSV := range allowedVersions {
-			if string(acceptableCSV) == ipCSVs[0] {
-				matchingCSV = true
+		policyTyped := policyv1beta1.OperatorPolicy{}
 
-				break
-			}
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(policy.Object, &policyTyped)
+		if err != nil {
+			continue
 		}
 
-		if matchingCSV {
-			approvedVersion = ipCSVs[0]
+		subTyped := operatorv1alpha1.Subscription{}
+
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(sub.Object, &subTyped)
+		if err != nil {
+			continue
+		}
 
-			approvableInstallPlans = append(approvableInstallPlans, installPlan)
+		approvedCSV := r.getApprovedCSV(&policyTyped, &subTyped, csvNames)
+		if approvedCSV != "" {
+			approvedCSVs.Insert(approvedCSV)
 		}
 	}
 
-	opLog.V(2).Info("Determined approvable InstallPlans", "count", len(approvableInstallPlans))
+	unapprovedCSVs := requiredCSVs.Difference(approvedCSVs).UnsortedList()
 
-	if len(approvableInstallPlans) != 1 {
-		changed := updateStatus(
-			policy,
-			installPlanUpgradeCond(complianceConfig, allUpgradeVersions, approvableInstallPlans),
-			relatedInstallPlans...,
-		)
+	sort.Strings(unapprovedCSVs)
 
-		return changed, nil
+	return unapprovedCSVs, nil
+}
+
+// getApprovedCSV returns the CSV in the passed in csvNames that can be approved. If no approval can take place, an
+// empty string is returned.
+func (r *OperatorPolicyReconciler) getApprovedCSV(
+	policy *policyv1beta1.OperatorPolicy, sub *operatorv1alpha1.Subscription, csvNames []string,
+) string {
+	// Only enforce policies can approve InstallPlans
+	if !policy.Spec.RemediationAction.IsEnforce() {
+		return ""
 	}
 
-	opLog.Info("Approving InstallPlan", "InstallPlanName", approvableInstallPlans[0].GetName(),
-		"InstallPlanNamespace", approvableInstallPlans[0].GetNamespace())
+	initialInstall := sub.Status.InstalledCSV == ""
+	autoUpgrade := policy.Spec.UpgradeApproval == "Automatic"
 
-	if err := unstructured.SetNestedField(approvableInstallPlans[0].Object, true, "spec", "approved"); err != nil {
-		return false, fmt.Errorf("error approving InstallPlan: %w", err)
+	// If the operator is already installed and isn't configured for automatic upgrades, then it can't approve
+	// InstallPlans.
+	if !autoUpgrade && !initialInstall {
+		return ""
 	}
 
-	if err := r.TargetClient.Update(ctx, &approvableInstallPlans[0]); err != nil {
-		return false, fmt.Errorf("error updating approved InstallPlan: %w", err)
+	allowedCSVs := make([]policyv1.NonEmptyString, 0, len(policy.Spec.Versions)+1)
+	allowedCSVs = append(allowedCSVs, policy.Spec.Versions...)
+
+	if sub.Spec.StartingCSV != "" {
+		allowedCSVs = append(allowedCSVs, policyv1.NonEmptyString(sub.Spec.StartingCSV))
+	}
+
+	// All versions for the operator are allowed if no version is specified
+	if len(policy.Spec.Versions) == 0 {
+		// currentCSV is the CSV related to the latest InstallPlan
+		if sub.Status.CurrentCSV != "" {
+			return sub.Status.CurrentCSV
+		}
```

Comment:
Is there a guarantee that `sub.Status.CurrentCSV` is in `csvNames`?

Reply:
It feels like there could be items in `csvNames` that "should" be approved by a policy with no versions specified - I thought we were considering some sort of name matching in this case. Is that somewhere else? Or is it not necessary?

Reply:
Here are the breadcrumbs:
https://github.com/operator-framework/operator-lifecycle-manager/blob/889ae50a66fe36a46a3c756a0ad7b481b73672f1/pkg/controller/operators/catalog/operator.go#L1291
https://github.com/operator-framework/operator-lifecycle-manager/blob/889ae50a66fe36a46a3c756a0ad7b481b73672f1/pkg/controller/registry/resolver/step_resolver.go#L123-L127
https://github.com/operator-framework/operator-lifecycle-manager/blob/889ae50a66fe36a46a3c756a0ad7b481b73672f1/pkg/controller/registry/resolver/step_resolver.go#L132-L136
https://github.com/operator-framework/operator-lifecycle-manager/blob/889ae50a66fe36a46a3c756a0ad7b481b73672f1/pkg/controller/registry/resolver/step_resolver.go#L197-L204
https://github.com/operator-framework/operator-lifecycle-manager/blob/889ae50a66fe36a46a3c756a0ad7b481b73672f1/pkg/controller/operators/catalog/operator.go#L1428-L1433

Reply:
@JustinKuli I'll revise my statement of "latest in the channel" but rather the latest that replaces the current version.